File: LeetCode/Medium/Search Suggestion System/Go/search_suggestion_system.go
package main

import "fmt"
import "sort"

type TrieNode struct {
	words    []string
	children []*TrieNode
}

func New() *TrieNode { return &TrieNode{words: make([]string, 0), children: make([]*TrieNode, 26)} }

func insert(root *TrieNode, word string) {
	cur := root
	for i := 0; i < len(word); i++ {
		ch := word[i] - 'a'
		if cur.children[ch] == nil {
			cur.children[ch] = New()
		}
		if len(cur.children[ch].words) < 3 {
			cur.children[ch].words = append(cur.children[ch].words, word)
		}
		cur = cur.children[ch]
	}
}

func search_word(str string, root *TrieNode) [][]string {
	words := [][]string{}
	curr := root
	for i := 0; i < len(str); i++ {
		ch := str[i] - 'a'
		if curr.children[ch] != nil {
			words = append(words, curr.children[ch].words)
			curr = curr.children[ch]
		} else {
			words = append(words, []string{})
			curr.children[ch] = New()
			curr = curr.children[ch]
		}
	}

	return words
}

func suggested_products(prod []string, search string) [][]string {
	sort.Strings(prod)
	tr := New()
	for i := 0; i < len(prod); i++ {
		insert(tr, prod[i])
	}

	return search_word(search, tr)
}

func main() {
	fmt.Println(suggested_products([]string{"mobile", "mouse", "moneypot", "monitor", "mousepad"}, "mouse"))
}

File: LeetCode/Medium/Search Suggestion System/Java/searchSuggestionSystem.java
class Node {
    Node[] children;
    List<String> word;
    public Node() {
        this.children = new Node[26];
        this.word = new ArrayList<>();
    }
}

class Solution {
    Node trie;

    private void insert(String word) {
        Node root = trie;
        for (int i=0; i<word.length(); i++) {
            int k = word.charAt(i)-'a';
            if (root.children[k] == null) root.children[k] = new Node();
            if (root.children[k].word.size() < 3) root.children[k].word.add(word);
            root = root.children[k];
        }
    }

    private List<List<String>> search(String word) {
        Node root = trie;
        List<List<String>> res = new ArrayList<>();
        for (int i=0; i<word.length(); i++) {
            int k = word.charAt(i)-'a';
            if (root.children[k] == null) root.children[k] = new Node();
            res.add(root.children[k].word);
            root = root.children[k];
        }
        return res;
    }

    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        trie = new Node();
        for (String prod : products) insert(prod);

        return search(searchWord);
    }
}
